### PR TITLE
CP-32446: support for CPUID features available only during migration

### PR DESCRIPTION
**Pull request by edwintorok**
See commit messages for details.

### CODE REVIEW

File: ocaml/xapi/cpuid_helpers.ml [hunk 1]
```diff
@@ -84,24 +85,25 @@ let is_strict_subset left right =
 
 (** Field definitions for checked string map access *)
 let features_t   = Map_check.pickler features_of_string string_of_features
-let features     = Map_check.field "features" features_t
-let features_pv  = Map_check.field "features_pv" features_t
-let features_hvm = Map_check.field "features_hvm" features_t
+let features = Map_check.field Xapi_globs.cpu_info_features_key features_t
```
Comment (rwdrich) [inline ocaml/xapi/cpuid_helpers.ml]: I think I prefer the older formatting given that most of the lines are the same. Helps to highlight the differences I think

File: ocaml/xapi/cpuid_helpers.ml [hunk 2]
```diff
@@ -84,24 +85,25 @@ let is_strict_subset left right =
...
+let features_hvm = Map_check.field Xapi_globs.cpu_info_features_hvm_key features_t
+let features_pv_host = Map_check.field Xapi_globs.cpu_info_features_pv_host_key features_t
+let features_hvm_host = Map_check.field Xapi_globs.cpu_info_features_hvm_host_key features_t
 let cpu_count    = Map_check.(field "cpu_count" int)
 let socket_count = Map_check.(field "socket_count" int)
 let vendor       = Map_check.(field "vendor" string)
 
-let get_flags_for_vm ~__context vm cpu_info =
-  let features_key =
+let get_flags_for_vm ~__context ?(apply_migration=false) vm cpu_info =
```
Comment (rwdrich) [inline ocaml/xapi/cpuid_helpers.ml]: Do we use this new parameter?

Comment (edwintorok) [reply ocaml/xapi/cpuid_helpers.ml]: No, deleted

File: ocaml/xapi/cpuid_helpers.ml (hunk 1 above)
Comment (edwintorok) [reply ocaml/xapi/cpuid_helpers.ml]: I've added some spaces to align as it was before

File: ocaml/xapi/create_misc.ml [hunk 3]
```diff
@@ -604,10 +604,8 @@ let create_pool_cpuinfo ~__context =
            else
              pool'
     with Not_found ->
-      (* If the host doesn't have all the keys we expect, assume that we
-         			   are in the middle of an RPU and it has not yet been upgraded, so
-         			   it should be ignored when calculating the pool level *)
-      pool
+      (* pre-Dundee? *)
+      failwith (Printf.sprintf "Host %s is missing required `features*` keys" (Ref.string_of hostref))
```
[the remark this answers is not in the document]
Comment (edwintorok) [reply ocaml/xapi/create_misc.ml]: Replaced with a warning.

File: ocaml/xapi/create_misc.ml [hunk 4]
```diff
@@ -583,23 +587,25 @@ let create_pool_cpuinfo ~__context =
...
+    getf ~default key host in
+
   let merge pool (hostref, host) =
     try
       pool
       |> setf vendor (getf vendor host)
       |> setf cpu_count ((getf cpu_count host) + (getf cpu_count pool))
       |> setf socket_count ((getf socket_count host) + (getf socket_count pool))
-      |> setf features_pv (Cpuid_helpers.intersect (getf features_pv host) (getf features_pv pool))
+      |> setf features_pv (Cpuid_helpers.intersect (getfdefault ~defaultf:features_pv features_pv_host host) (getf features_pv pool))
```
[the remark this answers is not in the document]
Comment (edwintorok) [reply ocaml/xapi/create_misc.ml]: The pool only has keys for pool leveling, not for migration. Although I could rename it

[the remark this answers is not in the document]
Comment (edwintorok) [reply ocaml/xapi/create_misc.ml]: I've now made `features_hvm` on the pool simply be an intersection of `features_hvm` from all hosts, and similarly for `features_hvm_host` (with a fallback to reading from `features_hvm`). And I've updated the reset-cpu-flags function to use `features_hvm_host` from the pool instead. It is still slightly odd to have something with host in its name at the pool level, but we could say that it is the intersection of the features_hvm from the hosts to make sense.

[the remark this answers is not in the document]
Comment (robhoes) [reply ocaml/xapi/create_misc.ml]: Thanks, I think that it is clearer now in `Cpuid_helpers`, where it is more explicit what is used for migration and what at boot time.

**Comment by edwintorok** [comment]
Squashed the fixups

**Pull request merged**